### ISSUE TITLE
TreeView getPathAtPos g_box_free exceptions

### ISSUE THREAD

**Issue by Dankr4d** [StefanSalewski/gintro#158]
When using `getPathAtPos` in `motion-notify-event` and moving the mouse inside the treeview, but bellow the rows/entries (blank area) for a while, glib throws `GLib-GObject-CRITICAL [...] g_boxed_free: assertion 'boxed != NULL' failed.` (after a few seconds). Tested this with gintro version 0.8.3 and then upgraded to 0.9.3, but the exceptions are still thrown.

Example code:
```nim
import gintro/[glib, gobject, gtk, gdk]
import gintro/gio except ListStore

var list: TreeView

proc typeTest(o: gobject.Object; s: string): bool =
  let gt = g_type_from_name(s)
  return g_type_check_instance_is_a(cast[ptr TypeInstance00](o.impl), gt).toBool

proc listStore(o: gobject.Object): gtk.ListStore =
  assert(typeTest(o, "GtkListStore"))
  cast[gtk.ListStore](o)

proc onMotionNotifyEvent(self: TreeView, event: EventMotion): bool =
  var x, y: cdouble
  if not event.getCoords(x, y):
    return
  var treePath: TreePath
  var column: TreeViewColumn
  var cellX, cellY: int
  if not self.getPathAtPos(x.int, y.int, treePath, column, cellX, cellY):
    return
  # [...]

proc initList(list: TreeView) =
  let renderer = newCellRendererText()
  let column = newTreeViewColumn()
  column.title = "List Item"
  column.packStart(renderer, true)
  column.addAttribute(renderer, "text", 0)
  discard list.appendColumn(column)
  let gtype = gStringGetType()
  let store = newListStore(1, cast[ptr GType]( unsafeaddr gtype))
  list.setModel(store)

proc appActivate(app: Application) =
  let
    window = newApplicationWindow(app)
    sw = newScrolledWindow()
  window.title = "List view"
  window.position = WindowPosition.center
  window.borderWidth = 10
  window.setSizeRequest(370, 270)
  list = newTreeView()
  sw.add(list)
  sw.setPolicy(PolicyType.automatic, PolicyType.automatic)
  sw.setShadowType(ShadowType.etchedIn)
  list.setHeadersVisible(false)
  window.add(sw)
  initList(list)
  let selection = getSelection(list)
  list.connect("motion-notify-event", onMotionNotifyEvent)
  block TEST_DATA:
    var
      val: Value
      iter: TreeIter
    let store = listStore(getModel(list))
    discard init(val, gStringGetType())
    for num in 1..5:
      setString(val, "Test " & $num)
      store.append(iter)
      store.setValue(iter, 0, val)
  showAll(window)

proc main =
  let app = newApplication("org.gtk.example")
  connect(app, "activate", appActivate)
  discard run(app)

main()
```

**Comment by StefanSalewski** [comment]
Thanks for reporting, I will try to investigate that issue soon.

I think I remember your name, you are the one who created a real app with gintro some years ago -- and I think you are still the only one.

**Comment by StefanSalewski** [comment]
I can verify that issue, but only with default refc GC. Was you able to get that messages with --gc:arc too?

And I hope you will be able to migrate to GTK4 soon, it is nearly available since one year now, so most people should have it.

**Comment by StefanSalewski** [comment]
Yes, seems to be a refc issue only. Adding a GC_fullcollect() shows the messages immediately:

```
proc onMotionNotifyEvent(self: TreeView, event: EventMotion): bool =
  GC_fullcollect()
  var x, y: cdouble
```

I have not really an idea yet, but seems to be easy, will try to fix it in the next days...

**Comment by Dankr4d** [comment]
> I think I remember your name, you are the one who created a real app with gintro some years ago [...]

I'm still developing the app (https://github.com/Dankr4d/BF2142Unlocker), whenever I've time/the mood.

> Was you able to get that messages with --gc:arc too?

I just tested it and didn't get the message. I'll continue using arc as garbage collector, if everything runs fine (need to test it).

> And I hope you will be able to migrate to GTK4 soon [...]

I'll try to upgrade when I've done my cleanups, because I need to test it for Linux/Windows 32bit and 64bit and it may take a while.

**Comment by StefanSalewski** [comment]
Yes the issue is obvious:

```
     output.writeLine("\nproc gBoxedFree" & $gRegisteredTypeInfoGetTypeName(info) & "*(self: " & mangleName(gBaseInfoGetName(
          info)) & ") =")
      if moduleNameSpace == "cairo": # we may manually free cairo objects!
        output.writeLine("  if not self.ignoreFinalizer and  self.impl != nil:")
        output.writeLine("    boxedFree(", getTypeProc, "(), ", "cast[ptr " & mangleName(gBaseInfoGetName(info)) & "00](self.impl))")
        output.writeLine("    self.impl = nil")
      else:
        output.writeLine("  if not self.ignoreFinalizer:")
        output.writeLine("    boxedFree(", getTypeProc, "(), ", "cast[ptr " & mangleName(gBaseInfoGetName(info)) & "00](self.impl))")
```

For the ARC destructors and for cairo module we do the test "self.impl != nil:" but not for the other cases. Fix is obvious, but I still try to remember why I leave it out at all and why it causes not issues more often. Will provide the fixed gintro package soon. (For a function like newButton() we would return nil when gtk_new_button() gives us nil, so we do not create a proxy object in this case at all, so no reason to check for "self.impl != nil:". I think that was the idea why I let that check out, but maybe thre are other cases...)

Generally it is a good idea to compile with --gc:arc when gintro is used, as old refc GC may show GC issues with a delay which makes debugging more difficult. And for cairo drawing ARC has the advantages that memory is released immediately. But I think there are some packages with do not work with ARC, and async may leak memory still.

**Comment by StefanSalewski** [comment]
Ah yes, easy again:

```nim
  var treePath: TreePath
  var column: TreeViewColumn
  var cellX, cellY: int
  if not self.getPathAtPos(x.int, y.int, treePath, column, cellX, cellY):
```

For this case treePath is not a function result, so we get a non nil proxy object with .impl == nil when gtk_tree_view_get_path_at_pos() returns nil. WE will see how we can fix this in best way -- "self.impl != nil:" check would be the simplest solution of course, but maybe we do not really need that test.

**Comment by Dankr4d** [comment]
> Generally it is a good idea to compile with --gc:arc when gintro is used [...]

Well, I stumble over the same problem as the last time when I used orc (same problem with arc). If there's for example an out of index exception in the event proc, the application crashes delayed or in my test by adding
```nim
  echo "BEFORE"
  var testSeq: seq[int] = @[]
  echo testSeq[1]
  echo "AFTER"
  ```
to the `onMotionNotifyEvent` proc, it doesn't even crash. Maybe the delayed crashes are only when connecting with builder (low level event procs):
```nim
builder.connectSignals(cast[pointer](nil))
```

I'm currently using nim version 1.4, but I'm going to create an issue for that (with some better test code) as it seems that this also happens on the high level event procs (this was my first test without using the builder).

So currently I need to stick to the old garbage collector.

**Comment by StefanSalewski** [comment]
The general problem of the traditional Nim refc GC was that garbage collection started delayed, so that early GTK code with bugs was running for a few minutes before the crash occurred. So before ARC we had to add GC_fullcollect() calls to enforce early crashes in case of bugs. With ARC debugging became easier, as destructors are called when a var goes out of scope, so we can see the crash immediately. And with ARC we can compile with -d:useMalloc and then run with valgrind, which is helpful too. For GTK valgrind gives some false positive memory leaks, as GTK itself does not release all its memory properly, but valgrind would detect illegal memory access and for GTKs memory leaks we can use valgrind white lists to ignore them.

**Comment by StefanSalewski** [comment]
OK, you may try the fix with

```
nimble uninstall gintro
nimble install gintro@#head
```

Fix is tiny:

```
$ diff ~/gintrotest/tests/gen.nim gen.nim 
3c3
< # v 0.9.4 2021-JUL-19
---
> # v 0.9.3 2021-JUN-29
1655,1656d1654
<         if gArgInfoMayBeNull(arg):
<           nilArgFix.add(h2)
1715d1712
<   var nilArgFix: seq[string]
2149c2146
<               if nilArgFix.len > 0 or pars.tempResFree != "": # ugly, do we have a better test for this case
---
>               if pars.tempResFree != "": # ugly, do we have a better test for this case
2190,2194d2186
<       for n in nilArgFix: # https://github.com/StefanSalewski/gintro/issues/158
<         methodBuffer.writeLine("  if $1 != nil and $1.impl == nil:" % n)
<         methodBuffer.writeLine("    $1.ignoreFinalizer = true" % n)
<         methodBuffer.writeLine("    $1 = nil" % n)
< 
4280c4272
< # 4280 lines
---
> # 4263 lines defaultParameters
```

Maybe not the best fix we can think of, as we still allocate the proxy object first, and later set it to nil when .impl is nil, and as we have to use the ignoreFinalizer field. Using a temporary variable instead of .impl first may be a better solution, but that would be a larger non trivial change with much room for new bugs.

Important is, that if .impl gets the value nil, we return also nil as value for the proxy object. This seems to work, as in

```
proc onMotionNotifyEvent(self: TreeView, event: EventMotion): bool =
  GC_fullcollect()
  var x, y: cdouble
  if not event.getCoords(x, y):
    return
  var treePath: TreePath
  var column: TreeViewColumn
  var cellX, cellY: int
  let res = self.getPathAtPos(x.int, y.int, treePath, column, cellX, cellY)
  if not res:
    echo treePath == nil, column == nil # should be nil here
  # [...]
```

The gintro examples seems to still compile, but I have not yet tested them all.

**Comment by Dankr4d** [comment]
The exceptions are gone; seems to work fine.
I've also tested it with a lowlevel event proc (connected with `builder`).
```nim
# [...]
var list: TreeView
# [...]
proc onMotionNotifyEvent(self: ptr TreeView00, event00: ptr EventMotion00): bool {.exportc, cdecl, dynlib.} =
  var event: EventMotion = new EventMotion
  event.impl = event00
  event.ignoreFinalizer = true

  var x, y: cdouble
  if not event.getCoords(x, y):
    return

  var treePath: TreePath
  var column: TreeViewColumn
  var cellX, cellY: int
  if not list.getPathAtPos(x.int, y.int, treePath, column, cellX, cellY):
    return
# [...]
```

Thank you!

**Comment by Dankr4d** [comment]
Can I close this issue or should i leave it open?

**Comment by StefanSalewski** [comment]
I think we can leave it open until this fix will make it into the next tagged version.

This fix alone does not justify shipping a 0.9.4 version, so to get this fix we have to install still with

nimble install gintro@#head

**Comment by StefanSalewski** [comment]
Should be fixed in v0.9.4 now.
Will close this issue, please reopen if necessary.
